in_toto/verifylib.py

```diff
@@ -1043,8 +1044,105 @@ def reduce_chain_links(chain_link_dict):
+    by an equivalent link object
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    chain_link_dict:
+            A dictionary of key-link pair with step names as keys. For each
+            step name, there are one or more keyids and corresponding
+            link objects.
+
+  <Exceptions>
+    raises an Exception if verification of the delegated step fails.
+
+  <Side Effects>
+    None.
 
-def in_toto_verify(layout_path, layout_key_paths):
+  """
+
+  for step_name, key_link_dict in six.iteritems(chain_link_dict):
+
+    for keyid, link in six.iteritems(key_link_dict):
+
+      # Extract keyid and link from the passed chain_link_dict
+      # Check if the link object is of type 'layout'
+      if link._type == "layout":
+        layout_key_dict = {}
+
+        for key, key_obj in six.iteritems(layout.keys):
+
+          # Extract the key object from the passed layout
+          if key == keyid:
+            # Extract the key object of the key corresponding
+            # to the link object
+            layout_key_dict[key] = key_obj
+
+        # Make a recursive call to in_toto_verify with the
+        # layout and the extracted key object
+        return_link = in_toto_verify(link, layout_key_dict)
+
+        # Replace the layout object in the passed chain_link_dict
+        # with the link file returned by in-toto-verify
+        key_link_dict[keyid] = return_link
+
+  return chain_link_dict
+
+def get_return_link_file(layout, reduced_chain_link_dict):
+  """
+  <Purpose>
+    Merges the materials of the first step (as mentioned in the layout)
+    and the products of the last step and returns a new link file which
+    defines the entire software supply chain
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    reduced_chain_link_dict:
+            A dictionary of step-link pair with step names as keys.
+
+  <Exceptions>
+    None.
+
+  <Side Effects>
+    None.
+
+  """
+
+  # Create a new link object to return
+  return_link = in_toto.models.link.Link()
+
+  for step_name, link in six.iteritems(reduced_chain_link_dict):
+    # Extract the step and its link from the passed chain_link_dict
+    # Check if this step is the first step in the layout
+    if step_name == layout.steps[0].name:
+      # Copy the materials to the return link
+      return_link.materials = link.materials
+
+    # Check if this step is the last step in the layout
+    if step_name == layout.steps[len(layout.steps) - 1].name:
+      # Copy the products and other properties
+      # to the return link
+      return_link._type = link._type
+      return_link.name = link.name
+      return_link.products = link.products
+      return_link.byproducts = link.byproducts
+      return_link.command = link.command
+      return_link.return_value = link.return_value
```

Comment:
I think it would be good to describe why certain values are take from the first and others from the last step. Either inline or in the function docstring.